daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/ReinterpretTest.scala

```diff
@@ -0,0 +1,203 @@
+              Some(Identifier(miniTestsPkgId, "ReinterpretTests:MySimple")),
+              ImmArray((Some[Name]("p"), ValueParty(party))),
+            )
+          ),
+          "",
+        )
+    )
+
+  val lookupContract = defaultContracts.get(_)
+  val lookupPackage = allPackages.get(_)
+  val lookupKey = { _: GlobalKeyWithMaintainers => None }
+
+  private val engine = Engine.DevEngine()
+
+  private def Top(xs: Shape*) = Shape.Top(xs.toList)
+  private def Exercise(xs: Shape*) = Shape.Exercise(xs.toList)
+  private def Rollback(xs: Shape*) = Shape.Rollback(xs.toList)
+  private def Create() = Shape.Create()
+
+  val submitters = Set(party)
+  val time = Time.Timestamp.now()
+  val seed = hash("ReinterpretTests")
+
+  private def reinterpretCommand(theCommand: Command): SubmittedTransaction = {
+    val res = engine
+      .reinterpret(
+        submitters,
+        theCommand,
+        Some(seed),
+        time,
+        time,
+      )
+      .consume(
+        lookupContract,
+        lookupPackage,
+        lookupKey,
+        VisibleByKey.fromSubmitters(submitters),
+      )
+    val Right((tx, _)) = res
+    tx
+  }
+
+  "Reinterpretation" should {
+
+    "be correct for a successful exercise command" in {
+      val choiceName = "MyHello"
+      val theCommand = {
+        val templateId = Identifier(miniTestsPkgId, "ReinterpretTests:MySimple")
+        val r = Identifier(miniTestsPkgId, s"ReinterpretTests:$choiceName")
+        val cid = toContractId("#ReinterpretTests:MySimple:1")
+        ExerciseCommand(templateId, cid, choiceName, ValueRecord(Some(r), ImmArray.empty))
+      }
+      val tx = reinterpretCommand(theCommand)
+      Shape.ofTransaction(tx.transaction) shouldBe Top(Exercise())
+    }
+
+    "be a rollback for an exercise command which throws" in {
+      val choiceName = "MyThrow"
+      val theCommand = {
+        val templateId = Identifier(miniTestsPkgId, "ReinterpretTests:MySimple")
+        val r = Identifier(miniTestsPkgId, s"ReinterpretTests:$choiceName")
+        val cid = toContractId("#ReinterpretTests:MySimple:1")
+        ExerciseCommand(templateId, cid, choiceName, ValueRecord(Some(r), ImmArray.empty))
+      }
+      val tx = reinterpretCommand(theCommand)
+      Shape.ofTransaction(tx.transaction) shouldBe Top(Rollback(Exercise()))
+    }
+
+    "be correct for a successful create comamnd" in {
+      val theCommand = {
+        val templateId = Identifier(miniTestsPkgId, "ReinterpretTests:MyEnsuring")
+        CreateCommand(
+          templateId,
+          ValueRecord(
+            Some(templateId),
+            ImmArray(
+              (Some[Ref.Name]("p"), ValueParty(party)),
+              (Some[Ref.Name]("v"), ValueInt64(42)), // ok (< 50)
+            ),
+          ),
+        )
+      }
+      val tx = reinterpretCommand(theCommand)
+      Shape.ofTransaction(tx.transaction) shouldBe Top(Create())
+    }
+
+    "not be a rollback for a create which fails an ensure" in {
```

Comment:
This is not the case I was trying to get at. I mean the following:
1. Exercise C
2. In that exercise create a contract with an ensure that returns `False` (so not one produced by the compiler, those throw so they are catchable).

Reply:
didn't do this, as we discussed